doc/hacking.rst

```diff
@@ -315,3 +315,161 @@ For automatically updated builds, you can install sphinx-autobuild, and run
+  particularly annoying violator of this is the "logic-less chain of
+  functions". For example:
+
+.. code:: ocaml
+
+   let foo t = bar t |> baz
+
+  If ``bar`` and ``baz`` are already public, then there's no need to add yet
+  another helper to save the caller a line of code.
+
+- Define bindings as close to their use site as possible. When they're far
+  apart, reading code requires scrolling and IDE tools to understand the code.
+
+Bad:
+
+.. code:: ocaml
+
+   let dir = .. in
+   (* 50 odd lines or so that don't use [dir] *)
+   f dir
+
+Good:
+
+.. code:: ocaml
+
+  let dir = .. in
+  f dir
+
+- A corollary to the previous guideline: keep the scope of bindings as small as
+  possible.
+
+Bad:
+
+.. code:: ocaml
+
+   let x1 = f foo in let x2 = f bar in
+   let y1 = g foo in let y2 = g bar in
+   let dx = x2 -. x1 in
+   let dy = y2 -. y1 in
+   dx^2 +. dy^2
+
+Good:
+
+.. code:: ocaml
+
+   let dx =
+     let x1 = f foo in let x2 = f bar in
+     x2 -. x1
+   in
+   let dy =
+     let y1 = g foo in let y2 = g bar in
+     y2 -. y1
+   in
+   dx^2 +. dy^2
+
+- Prefer ``Code_error.raise`` instead of ``assert false``. The reader often has
+  no idea what invariant is broken by the ``assert false``. Kindly describe it
+  to the reader in the error message.
+
+- Avoid meangingless names like `x`, `a`, `b`, `f`. Try to find a more
+  descriptive name or just inline it altogether.
+
+- If a module ``Foo`` has a module type ``Foo.S`` and you'd like to avoid
+  repeating its definition in the implementation and the signature, introduce
+  an .ml only module ``Foo_intf`` and write the ``S`` only once in there.
+
+- Instead of introducing a type ``foo``, consider introducing a module ``Foo``
+  with a type ``t``. This is often a place to put functions related to ``foo``.
+
+Subjective Style Points
+-----------------------
+
+There's some stylistic decisions we made that don't have logical justification
+and are basically a matter of taste. Nevertheless, it's useful to follow them
+to keep the code consistent.
+
+- Match patterns should be sorted by the length of their RHS when possible.
+  Keep the shorter clauses near the top.
+
+- If a module ``Foo`` defines a type ``t``, all functions that take ``t`` in
+  this module should have ``t`` as their first argument. This is the "t comes
+  first" rule.
+
+- Do not write inverted if-else expressions. 
+
+Bad:
+
+.. code:: ocaml
+
+   (* try reading this out loud without short circuiting your brain *)
+   if not x then foo else bar
+
+Good:
+
+.. code:: ocaml
+
+   if x then bar else foo
+
+- We prefer snake_casing identifiers. This includes the names of modules and
+  module types.
+
+- Avoid qualifying constructors, record fields. Instead, add type annotations to the
+  type being matched on or being constructed. E.g.
+
+Bad:
+
+.. code:: ocaml
+
+   let foo = Command.Args.S []
+
+Good:
+
+.. code:: ocaml
+
+   let (foo : _ Command.Args.t) = S []
```

Comment:
I'd add something about avoiding optional arguments. It's another good rule we already apply.